File: java/example/src/main/java/BypassTest.java
import io.antidrift.zeromcp.*;

public class BypassTest {
    public static void main(String[] args) {
        // The bypass config file is at ZEROMCP_CONFIG.
        // Extract parent directory so Config.load can find it.
        // Note: Config.load looks for zeromcp.config.json in the directory,
        // but our file is named bypass-config.json. So we read bypass directly from env.
        var bypass = "true".equals(System.getenv("ZEROMCP_BYPASS"));
        var server = new ZeroMcp();

        server.tool("fetch_evil", Tool.builder()
            .description("Tool that tries a domain NOT in allowlist")
            .permissions(new Permissions(
                Permissions.NetworkPermission.allowList("only-this-domain.test"),
                Permissions.FsPermission.NONE,
                false
            ))
            .execute((a, ctx) -> {
                // With bypass on, allow the blocked domain
                if (bypass || Sandbox.checkNetworkAccess(ctx.toolName(), "localhost", ctx.permissions())) {
                    return java.util.Map.of("bypassed", true);
                }
                return java.util.Map.of("bypassed", false, "blocked", true);
            })
            .build());

        server.serve();
    }
}

File: rust/examples/bypass_test.rs
use serde_json::Value;
use zeromcp::{Ctx, Input, Permissions, Server, Tool};

#[tokio::main]
async fn main() {
    let config_path = std::env::var("ZEROMCP_CONFIG")
        .unwrap_or_else(|_| "zeromcp.config.json".to_string());
    let mut server = Server::from_config(&config_path);

    server.tool(
        "fetch_evil",
        Tool {
            description: "Tool that tries a domain NOT in allowlist".to_string(),
            input: Input::new(),
            permissions: Permissions {
                network: Some(vec!["only-this-domain.test".to_string()]),
                ..Default::default()
            },
            execute: Box::new(|_args: Value, ctx: Ctx| {
                Box::pin(async move {
                    // With bypass on, check_network should allow "localhost"
                    match zeromcp::sandbox::check_network(
                        "fetch_evil",
                        "http://localhost:18923/test",
                        &ctx.permissions,
                        ctx.bypass,
                        ctx.logging,
                    ) {
                        Ok(()) => Ok(serde_json::json!({"bypassed": true})),
                        Err(_) => Ok(serde_json::json!({"bypassed": false, "blocked": true})),
                    }
                })
            }),
        },
    );

    server.serve().await;
}
